Add unique_id for ZoneMinder cameras

nabbi opened the pull request:
This generates unique_id for ZoneMinder camera discovery as:
 "zoneminder-<ZM_monitor_name>-mid<ZM_monitor_id>"


<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Existing ZoneMinder deployments will import cameras as new Entity ID's.

This behavior occurs because previous ZM cameras were created without an
internal unique_id. This change was made to allow additional configuration to be
set by users from the HA UI.

Any configuration referencing those previous names will need to be remapped,
or you can re-attempt the discovery process by purging the previous non-uniqiue_id
entities and restarting HA to continue using the original default names.

1. [Backup Home Assistant](https://www.home-assistant.io/integrations/backup/)
2. Search "camera" under Settings > Devices & Service > Entities
3. Remove Selected, the old non-connected camera entities
4. Restart HA


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fixes #107208

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->
Unless someone has a trick when migrating from unique_id null to string,
I'm listing this as breaking change.

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #107208

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

nabbi committed:
unique_id for ZoneMinder cameras

This generates unique_id as:
 "zoneminder-<ZM_monitor_name>-mid<ZM_monitor_id>"

Closes #107208

Signed-off-by: Nic Boet <nic@boet.cc>

gjohansson-ST commented on homeassistant/components/zoneminder/camera.py:
```diff
@@ -56,6 +56,9 @@ def __init__(self, monitor: Monitor, verify_ssl: bool) -> None:
         self._monitor = monitor
 
+        zmname = monitor.name.casefold().replace(" ", "")
+        self._attr_unique_id = f"{ZONEMINDER_DOMAIN}-{zmname}-mid{monitor.id}"
```
Why isn't it sufficient with only `monitor.id`

nabbi replied:
In this context, `monitor.id` contains the zoneminder camera monitor id. While this is a unique integer within the ZM database tables, the counter increments starting at `1`

i.e. a unique_id value of `zoneminder-loadingdock-mid3` is more distinct than  `3`

gjohansson-ST replied:
It needs to be unique within the domain and platform so I still then unless more entities can have the same id it's sufficient to be unique?

There is no point in adding more to the unique id than to establish its unique.

rohankapoorcom replied:
You could be connected to multiple Zoneminder servers though. Each server would reuse the id numbers starting from `1`.

nabbi replied:
> You could be connected to multiple Zoneminder servers though. Each server would reuse the id numbers starting from `1`.

Agreed. Ideally combining the zm server hostname and mid would be the unique tuple for identifying the specific camera on a specific zm server. Yet hostnames are deemed an [Unacceptable sources for a unique ID](https://developers.home-assistant.io/docs/entity_registry_index/#unacceptable-sources-for-a-unique-id) so incorporating the monitor name should help minimize conflicts.

rohankapoorcom replied:
The better approach may be to switch to config flow with config entries for setup first and then utilizing the config entry id + monitor id to generate the unique id.

nabbi replied:
Another question I am curious about is with having the ZoneMinder component import MJpegCamera for creating device entities as `camera.loadingdock`. If we do go with the generic integer `3`, would this not leave us open to other unique_id conflicts within the `camera` domain that zoneminder does not control ? Manual camera adds? other video platform integrations? 
This was mainly why I added the `zoneminder` prefix to the string as I assumed we need to keep our camera uniquely referenced.

nabbi replied:
[07d103c](https://github.com/home-assistant/core/pull/107704/commits/07d103cc4a8688ea76b9bd70e2f3ab76844cf0fb) just passes a string to `MjpegCamera` which sets the `unqiue_id` 
Hopefully this better illustrates my last comment. Yet doesn't resolve the concerns raised.

gjohansson-ST replied:
It's not about minimizing conflicts. It has to be truly unique so it's not in this case.

I suggest you work out a config entry or something that can be used to make it unique.
Going to put this PR to draft again, don't undraft unless it's fixed to make it unique. Thanks 👍

nabbi committed:
refactor

this doesn't resolve the concerns in #107704

clarifying that homeassistant.components.mjpeg MjpegCamera
creates the "camera." entity for "zoneminder" platform

Signed-off-by: Nic Boet <nic@boet.cc>
